### PR TITLE
fix(Checkbox): prevent click propagation from the input element

### LINKED ISSUE #3433: Checkbox: onClick fires parent onClick twice
## Bug Report
I just upgraded to v0.85 (from 0.82.5) seeing this release is supposed to fix this issue, https://github.com/Semantic-Org/Semantic-UI-React/pull/3351. My issue might be slightly different though.

I have the following subset of code:
```
         <li
            onClick={this.liClickHandler}
          >
            <div className="border--left" />
            <div className="infoIconContainer">
              <Popup trigger={<Icon name="info circle" />} content={node.data.tooltipText} inverted />
            </div>
            <div className="checkboxContainer">
              <Checkbox
                id={node.id}
                checked={node.data.active}
                indeterminate={node.data.indeterminate}
                radio={treeType === 'Radio'}
                onClick={e => e.stopPropagation()}
              />...
```

When the user clicks the checkbox, which is a radio button, it runs the `li`'s `onClick` event twice. If the user clicks the `li` it runs once. 

It appears to only do this when I specify the `id` property. This can be validated in the test case below
 
### Steps
Use the checkbox with an ID specified and a parent click event and note it fires twice. It fires twice for radio and checkboxes.

Put onclick on the checkbox and it only fires once.

### Expected Result
it fires once

### Actual Result
it fires twice. In debugging the `this.liClickHandler`, I notice that the `event` object is slightly different the 2nd time it calls.

First time, `event.target` is `label`, the second time it's the `input`. Appears both controls are pushing the click.

### Version
    "semantic-ui-css": "^2.4.1",
    "semantic-ui-react": "^0.85.0",

### Testcase
https://codesandbox.io/s/88zzv64rn8

### LINKED ISSUE #3440: Checkbox: no longer prevents default after updating to version 0.85.0
## Bug Report
Hi, we are using a Card component with a toggle Checkbox in its header. The Card component is also used as a link (from react-router-dom), so we need to prevent the link from firing when we click the toggle. Everything was working fine until we updated to version 0.85.0 from 0.83.0.

### Steps
Use a card component (as a link) with a checkbox in it and try preventing default when you click the checkbox to prevent the link from firing.

### Expected Result
Checkbox is checked without being redirected to the link.

### Actual Result
Prevent default doesn't work anymore so you are redirected to the link.

### Version
"semantic-ui": "^2.4.2"
"semantic-ui-react": "^0.85.0"

### Testcase
I created an example without a link which uses stopPropagation instead but the described behavior can still be observed if switched between the two versions. I wonder if this could be caused by this PR: https://github.com/Semantic-Org/Semantic-UI-React/pull/3351

https://codesandbox.io/s/0oxky24lyp

EDIT: Here is an example with a link: https://codesandbox.io/s/2zxzr1jw5p

Thanks for any help!

### PR DESCRIPTION
**Pull request by Fabianopb**
Initially the click was being fired from both `label` and `input`
checkbox-antes

Now it only comes from `label`
checkbox-depois

(Closes #3433)
(Closes #3440)

### CODE REVIEW

**Comment by layershifter** [comment]
Actually we should check:
- only one `click` was propagated with `id` on `label`
- only one `click` was propagated with `id` on `input`
- only one `click` was propagated without `id` on `input`
- only one `click` was propagated without `id` on `label`

**Comment by Fabianopb** [comment]
I can check that later, but just to confirm, the `id` goes directly on the `Checkbox`, right? And then it's passed to the `input` and `label` within the component.

So if we're testing
```jsx
mount(
  <div role='presentation' onClick={onClick}>
    <Checkbox id='foo' />
  </div>,
)
```

We could also test
```jsx
mount(
  <div role='presentation' onClick={onClick}>
    <Checkbox />
  </div>,
)
```

Then it's all or nothing with the id, right?

**Comment by layershifter** [comment]
> the `id` goes directly on the `Checkbox`, right?

Absolutely

```jsx
const wrapperWithId = mount(
  <div role='presentation' onClick={onClick}>
    <Checkbox id='foo' />
  </div>,
)
const wrapper = mount(
  <div role='presentation' onClick={onClick}>
    <Checkbox />
  </div>,
)

wrapperWithId.find('input').click() // 1
wrapperWithId.find('label').click() // 2
wrapper.find('input').click() // 3
wrapper.find('label').click() // 4
```

**Comment by Fabianopb** [comment]
Oh that's right, we set the `id` in the `Checkbox`, but then we can simulate the click both on the `input` and on the `label`. I'll check that later!

**Comment by Fabianopb** [comment]
Tests updated @layershifter !
`inputs` don't propagate click because there's *always* a `label` on top of them, so the click event is always propagated from `label`.

File: test/specs/modules/Checkbox/Checkbox-test.js [hunk 1]
```diff
@@ -201,6 +201,50 @@ describe('Checkbox', () => {
...
+    assertions.forEach((assertion) => {
+      it(assertion.description, () => {
+        const onClick = sandbox.spy()
+        wrapperMount(
+          <div role='presentation' onClick={onClick}>
+            <Checkbox id={assertion.id} />
+          </div>,
+        )
+        const target = document.querySelector(`.ui.checkbox ${assertion.target}`)
+        domEvent.click(target)
```
Comment (layershifter) [inline test/specs/modules/Checkbox/Checkbox-test.js]:
```diff
-const target = document.querySelector(`.ui.checkbox ${assertion.target}`)
-domEvent.click(target)
+domEvent.click(`.ui.checkbox ${assertion.target}`)

This should be a valid change 😺

File: test/specs/modules/Checkbox/Checkbox-test.js [hunk 2]
```diff
@@ -201,6 +201,50 @@ describe('Checkbox', () => {
...
+        if (assertion.propagates) onClick.should.have.been.calledOnce()
+        else onClick.should.not.have.been.called()
```
Comment (layershifter) [inline test/specs/modules/Checkbox/Checkbox-test.js]:
```diff
-if (assertion.propagates) onClick.should.have.been.calledOnce()
-else onClick.should.not.have.been.called()
+onClick.should.have.callCount(Number(assertion.propagates))
```

We can test the call count directly, https://github.com/domenic/sinon-chai

File: test/specs/modules/Checkbox/Checkbox-test.js [hunk 3]
```diff
@@ -201,6 +201,50 @@ describe('Checkbox', () => {
     })
   })
 
+  describe('click propagation', () => {
+    const assertions = [
```
Comment (layershifter) [inline test/specs/modules/Checkbox/Checkbox-test.js]:
```suggestion
    const assertMatrix= [
```

File: src/modules/Checkbox/Checkbox.js [hunk 4]
```diff
@@ -258,6 +258,7 @@ export default class Checkbox extends Component {
           tabIndex={this.computeTabIndex()}
           type={type}
           value={value}
+          onClick={e => e.stopPropagation()}
```
Comment (layershifter) [inline src/modules/Checkbox/Checkbox.js]: Can we create a `handleInputClick` method and make a comment there with a link to the matching issue?

Comment (layershifter) [reply src/modules/Checkbox/Checkbox.js]: And sort props in alphabetic order ❤️

**Comment by layershifter** [comment]
@Fabianopb I made some comments, can you please address them? Nice job 👍

**Comment by Fabianopb** [comment]
Awesome @layershifter thanks for the comments, I'll try address them later today!

File: src/modules/Checkbox/Checkbox.js (hunk 4 above)
Comment (Fabianopb) [reply src/modules/Checkbox/Checkbox.js]: Done!

File: test/specs/modules/Checkbox/Checkbox-test.js (hunk 3 above)
Comment (Fabianopb) [reply test/specs/modules/Checkbox/Checkbox-test.js]: Sure, I'll change that

File: test/specs/modules/Checkbox/Checkbox-test.js (hunk 1 above)
Comment (Fabianopb) [reply test/specs/modules/Checkbox/Checkbox-test.js]: Good to know!

File: test/specs/modules/Checkbox/Checkbox-test.js (hunk 2 above)
Comment (Fabianopb) [reply test/specs/modules/Checkbox/Checkbox-test.js]: Nice idea!

**Review by Fabianopb — COMMENTED** [review]
@layershifter I just pushed the changes!

**Comment by layershifter** [comment]
@Fabianopb I added changes that I picked from @levithomason branch 😸 Can you please help to test these changes? Did I miss something? I.e. double clicks, event propagation, right mouse click?

File: test/specs/modules/Checkbox/Checkbox-test.js [hunk 5]
```diff
@@ -424,40 +387,77 @@ describe('Checkbox', () => {
...
+        },
+        target: 'label',
+      },
+      {
+        description: 'click on input: fires on mouse click',
+        events: {
+          label: ['mouseup', 'click'],
+          input: ['click'],
+        },
         target: 'label',
```
Comment (Fabianopb) [inline test/specs/modules/Checkbox/Checkbox-test.js]: This `target` property doesn't seem to be used anymore, is that so?

**Review by Fabianopb — COMMENTED** [review]
@layershifter that looks good to me, just one detail I got from re-testing the behavior manually:

* without "id" the parent listens for a click coming necessarily from the `label`
* with "id" the parent listens for a click coming necessarily from the `input`

Is that ok/intended?

File: test/specs/modules/Checkbox/Checkbox-test.js (hunk 5 above)
Comment (layershifter) [reply test/specs/modules/Checkbox/Checkbox-test.js]: Nice catch 👍

**Comment by layershifter** [comment]
> Is that ok/intended?

For me it looks fine, I don't see a big issue. I've tested it again, works like a charm 💎 

Thank you!

**Comment by levithomason** [comment]
@Fabianopb You've contributed here before and you've shown care and initiative for ensuring SUIR becomes better. We appreciate this and want to enable you to do more. I've added you as a collaborator. You're now able to manage issues and PRs as well as create branches here in the main repo. You no longer need to work from a fork.  Merging to master is the only reserved permission.

We are very select about who we add and when, so know that we trust your abilities and your judgment. Please feel free to weigh in on any issues and PRs that you have an opinion about. You have a share of ownership and say here now.

Once you accept the invite (see your email), you might want to show the Semantic Org badge publicly on your profile. To do that, navigate to the "people" page and change the "private" dropdown to "public" for your user in the table. Here's a [direct link](https://github.com/orgs/Semantic-Org/people?utf8=%E2%9C%93&query=Fabianopb) to your user in the table.

/cc
@Semantic-Org/react-collaborators
@Semantic-Org/react-maintainers

Please welcome @Fabianopb to our ranks 😄 🎉

**Comment by levithomason** [comment]
Released in `semantic-ui-react@0.86.0`.

**Pull request merged**